File: samples/python/securefusion_anchor.py
"""
SecureFusion v1.0 (SF1) anchor producer -- Python sample.

Builds the three-memo XRPL transaction payload for a video event:
  - SF1.bundle (50-byte binary header: hex-encoded)
  - SF1.event  (canonical manifest JSON: hex-encoded UTF-8)
  - SF1.sig    (Ed25519 signature over SF1.bundle || SF1.event: hex-encoded)

Submission to XRPL is left to a library such as `xrpl-py`:
    pip install xrpl-py

This sample has zero third-party dependencies for hashing / canonicalisation.
For Ed25519 signing, install one of:
    pip install cryptography
    pip install pynacl

Run:
    python securefusion_anchor.py

Expected output for the bundled example:
    bundleHash: e1b8a2206cf8c53754c392f8e6b7aad25972e6349a70bd7c4cecde424caf0f05
"""

from __future__ import annotations

import hashlib
import json
import sys
import uuid
from pathlib import Path
from typing import Any


# --------------------------------------------------------------------------
# Canonical JSON serialisation (RFC 8785-compatible subset).
# --------------------------------------------------------------------------

def canonicalise(value: Any) -> bytes:
    """Return the canonical UTF-8 byte representation of `value`.

    Rules:
      - object keys sorted lexicographically
      - no insignificant whitespace
      - integers without decimal point
      - boolean/null lowercase
    """
    return _serialise(value).encode("utf-8")


def _serialise(v: Any) -> str:
    if v is None:
        return "null"
    if v is True:
        return "true"
    if v is False:
        return "false"
    if isinstance(v, (int,)) and not isinstance(v, bool):
        return str(v)
    if isinstance(v, float):
        if v != v or v in (float("inf"), float("-inf")):
            raise ValueError("Non-finite numbers are not permitted in canonical JSON")
        # Python's repr produces a round-trippable shortest form.
        if v.is_integer():
            return str(int(v))
        return repr(v)
    if isinstance(v, str):
        return _serialise_string(v)
    if isinstance(v, list):
        return "[" + ",".join(_serialise(item) for item in v) + "]"
    if isinstance(v, dict):
        keys = sorted(v.keys())
        return "{" + ",".join(_serialise_string(k) + ":" + _serialise(v[k]) for k in keys) + "}"
    raise TypeError(f"Cannot canonicalise {type(v).__name__}")


def _serialise_string(s: str) -> str:
    out = ['"']
    for ch in s:
        c = ord(ch)
        if c == 0x22:
            out.append('\\"')
        elif c == 0x5C:
            out.append("\\\\")
        elif c == 0x08:
            out.append("\\b")
        elif c == 0x0C:
            out.append("\\f")
        elif c == 0x0A:
            out.append("\\n")
        elif c == 0x0D:
            out.append("\\r")
        elif c == 0x09:
            out.append("\\t")
        elif c < 0x20:
            out.append(f"\\u{c:04x}")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


# --------------------------------------------------------------------------
# SHA-256 helpers.
# --------------------------------------------------------------------------

def sha256_hex(data: bytes | str) -> str:
    """Return the lowercase hex SHA-256 of bytes or a UTF-8 string."""
    if isinstance(data, str):
        data = data.encode("utf-8")
    return hashlib.sha256(data).hexdigest()


def sha256_bytes(data: bytes) -> bytes:
    """Return the raw SHA-256 digest."""
    return hashlib.sha256(data).digest()


def sha256_file(path: str | Path) -> str:
    """Stream a file and return its lowercase hex SHA-256."""
    h = hashlib.sha256()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            h.update(chunk)
    return h.hexdigest()


# --------------------------------------------------------------------------
# SF1.bundle binary header (50 bytes).
# --------------------------------------------------------------------------

INGEST_SOURCE_CODES: dict[str, int] = {
    # SecureFusion v1: single-value enum on the wire, single source-code byte.
    # Per-upstream-provider distinctions are not exposed publicly. v2 may extend.
    "fleetfusion": 1,
}


def encode_bundle_memo(*, bundle_hash: str, event_id: str, ingest_source: str, channel_count: int) -> bytes:
    """Build the 50-byte SF1.bundle binary header.

    Layout:
      [0:32]   bundleHash
      [32:48]  eventId (raw 16-byte UUID, big-endian)
      [48]     ingestSource code
      [49]     channelCount
    """
    if len(bundle_hash) != 64:
        raise ValueError("bundleHash must be 64 hex chars (32 bytes)")
    src_code = INGEST_SOURCE_CODES.get(ingest_source)
    if src_code is None:
        raise ValueError(f"Unknown ingestSource: {ingest_source}")
    if not 1 <= channel_count <= 255:
        raise ValueError("channelCount must be 1..255")

    out = bytearray(50)
    out[0:32] = bytes.fromhex(bundle_hash)
    # FleetFusion writes the eventId via .NET Guid.ToByteArray(), which is
    # little-endian for the first 3 fields. Python's uuid.UUID.bytes_le
    # matches that exact wire format.
    out[32:48] = uuid.UUID(event_id).bytes_le
    out[48] = src_code
    out[49] = channel_count
    return bytes(out)


# --------------------------------------------------------------------------
# Memo construction for the XRPL Payment transaction.
# --------------------------------------------------------------------------

def _str_to_hex(s: str) -> str:
    return s.encode("utf-8").hex().upper()


def build_memos(*, bundle_bytes: bytes, event_bytes: bytes, signature: bytes) -> list[dict]:
    """Return the Memos array suitable for an XRPL Payment transaction.

    Each entry is a dict with a single 'Memo' key containing
    MemoType, MemoFormat, MemoData (all hex-encoded uppercase),
    matching the format expected by xrpl-py and rippled.
    """
    if len(bundle_bytes) != 50:
        raise ValueError("SF1.bundle must be 50 bytes")
    if len(signature) != 64:
        raise ValueError("Ed25519 signature must be 64 bytes")

    return [
        {
            "Memo": {
                "MemoType": _str_to_hex("SF1.bundle"),
                "MemoFormat": _str_to_hex("application/octet-stream"),
                "MemoData": bundle_bytes.hex().upper(),
            }
        },
        {
            "Memo": {
                "MemoType": _str_to_hex("SF1.event"),
                "MemoFormat": _str_to_hex("application/json"),
                "MemoData": event_bytes.hex().upper(),
            }
        },
        {
            "Memo": {
                "MemoType": _str_to_hex("SF1.sig"),
                "MemoFormat": _str_to_hex("application/octet-stream"),
                "MemoData": signature.hex().upper(),
            }
        },
    ]


# --------------------------------------------------------------------------
# Ed25519 signing via cryptography or PyNaCl.
# --------------------------------------------------------------------------

def ed25519_sign(message: bytes, secret_key_32: bytes) -> bytes:
    """Sign `message` with a 32-byte Ed25519 seed.

    Tries the `cryptography` library first, then PyNaCl. If neither is
    installed, raises a clear error.
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
        key = Ed25519PrivateKey.from_private_bytes(secret_key_32)
        return key.sign(message)
    except ImportError:
        pass
    try:
        from nacl.signing import SigningKey
        return bytes(SigningKey(secret_key_32).sign(message).signature)
    except ImportError:
        pass
    raise RuntimeError(
        "Neither 'cryptography' nor 'pynacl' is installed. "
        "Install one with: pip install cryptography  (or)  pip install pynacl"
    )


# --------------------------------------------------------------------------
# End-to-end builder.
# --------------------------------------------------------------------------

def build_anchor_payload(manifest: dict, app_signing_key_32: bytes | None = None) -> dict:
    """Build the full SecureFusion anchor payload for an event.

    Returns a dict with:
        bundleHash:   the canonical SHA-256 over the manifest
        bundleBytes:  raw 50-byte SF1.bundle header (bytes)
        eventBytes:   canonical UTF-8 JSON of the manifest (bytes)
        signature:    Ed25519 signature over bundleBytes || eventBytes
        memos:        XRPL Memos array ready to attach to a Payment

    If app_signing_key_32 is None, signature is 64 zero bytes (for testing
    canonicalisation only -- never submit such a transaction).
    """
    event_bytes = canonicalise(manifest)
    bundle_hash = sha256_hex(event_bytes)
    bundle_bytes = encode_bundle_memo(
        bundle_hash=bundle_hash,
        event_id=manifest["eventId"],
        ingest_source=manifest["ingestSource"],
        channel_count=len(manifest["channels"]),
    )

    sig_input = bundle_bytes + event_bytes
    if app_signing_key_32 is None:
        signature = bytes(64)
    else:
        signature = ed25519_sign(sig_input, app_signing_key_32)

    memos = build_memos(
        bundle_bytes=bundle_bytes,
        event_bytes=event_bytes,
        signature=signature,
    )

    return {
        "bundleHash": bundle_hash,
        "bundleBytes": bundle_bytes,
        "eventBytes": event_bytes,
        "signature": signature,
        "memos": memos,
    }


# --------------------------------------------------------------------------
# Demonstration entry point.
# --------------------------------------------------------------------------

EXAMPLES_DIR = Path(__file__).resolve().parent.parent.parent / "examples"

EXPECTED_BUNDLE_HASHES = {
    # Mirror of examples/test-vectors.json. Update both files together.
    "single-channel-event.json": "e1b8a2206cf8c53754c392f8e6b7aad25972e6349a70bd7c4cecde424caf0f05",
    "four-channel-event.json": "8e4a41ebcc87183ca42f023409713ac2316f37c480bae00582536759543936fb",
}


def main() -> int:
    print("SecureFusion v1.0 -- Python anchor producer sample")
    print("=" * 60)

    all_ok = True
    for filename, expected in EXPECTED_BUNDLE_HASHES.items():
        path = EXAMPLES_DIR / filename
        with path.open("r", encoding="utf-8") as f:
            manifest = json.load(f)

        payload = build_anchor_payload(manifest)
        ok = payload["bundleHash"] == expected
        all_ok &= ok

        print(f"\n  Manifest:    {filename}")
        print(f"  bundleHash:  {payload['bundleHash']}")
        print(f"  expected:    {expected}")
        print(f"  match:       {'[OK]' if ok else '[FAIL]'}")
        print(f"  channels:    {len(manifest['channels'])}")
        print(f"  memos:       {len(payload['memos'])}")
        print(f"  bundle hex:  {payload['bundleBytes'].hex()[:64]}...")

    print()
    if all_ok:
        print("[OK] All test vectors match.")
        return 0
    print("[FAIL] One or more test vectors did not match.")
    return 1


def self_test() -> int:
    """Minimal smoke test that runs to completion under any code page.

    Reads the bundled examples, recomputes the bundleHash, and compares to
    the published EXPECTED_BUNDLE_HASHES. Exits non-zero if any vector
    diverges. ASCII output only -- safe under cp1252 / cp437 PowerShell.
    """
    print("SecureFusion -- Python self-test")
    failed = []
    for filename, expected in EXPECTED_BUNDLE_HASHES.items():
        path = EXAMPLES_DIR / filename
        with path.open("r", encoding="utf-8") as f:
            manifest = json.load(f)
        got = build_anchor_payload(manifest)["bundleHash"]
        status = "[OK]" if got == expected else "[FAIL]"
        print(f"  {status} {filename}: {got}")
        if got != expected:
            failed.append((filename, expected, got))
    if failed:
        print(f"[FAIL] self-test: {len(failed)} vector(s) diverged")
        return 1
    print("[OK] self-test passed")
    return 0


if __name__ == "__main__":
    if "--self-test" in sys.argv[1:]:
        sys.exit(self_test())
    sys.exit(main())

File: samples/java/SecureFusionAnchor.java
/*
 * SecureFusion v1.0 (SF1) anchor producer - Java sample.
 *
 * Mirrors FleetAssistant.SecureFusion.SecureManifestBuilder + the C# reference
 * sample byte-for-byte:
 *   - Hand-rolled canonicaliser (alpha-sort keys, no whitespace, integer-only
 *     numerics, omit-when-empty for tenantId / signerKeyId).
 *   - ingestSource is the literal string "fleetfusion" (single-value enum in v1).
 *   - Manual timestamp formatting -- NEVER Instant.toString(), which strips
 *     trailing zero-millisecond digits and would diverge from FleetFusion's
 *     "yyyy-MM-dd'T'HH:mm:ss.fff'Z'" (3-digit ms, suffix Z).
 *   - SF1.bundle layout: bundleHash(32) || eventIdGuid(16) || sourceCode(1) ||
 *     channelCount(1), 50 bytes total. eventIdGuid uses .NET Guid.ToByteArray()
 *     byte order (mixed-endian: Data1/2/3 little-endian, Data4 verbatim) to
 *     match FleetFusion. The default Java UUID.getMostSignificantBits() byte
 *     order is RFC 4122 big-endian and would NOT match -- toDotNetGuidBytes()
 *     does the byte-swap.
 *   - SF1.sig: Ed25519 signature (BouncyCastle) over (bundleBytes || eventBytes)
 *     using the deterministic test seed (32 bytes 0x00..0x1F per
 *     joint-plan-final §D11).
 *
 * Build & run (Java 17+):
 *     cd samples/java
 *     # If using BouncyCastle:
 *     #   download bcprov-jdk18on-1.78.jar into ./lib/
 *     javac -cp "lib/*" SecureFusionAnchor.java
 *     java  -cp ".;lib/*" SecureFusionAnchor    (Windows)
 *     java  -cp ".:lib/*" SecureFusionAnchor    (Linux/macOS)
 *
 * Or via Maven -- a pom.xml in this directory wires bcprov-jdk18on so
 * `mvn compile exec:java -Dexec.mainClass=SecureFusionAnchor` does the
 * right thing.
 *
 * Expected bundleHashes for the bundled examples:
 *     single-channel:  e1b8a2206cf8c53754c392f8e6b7aad25972e6349a70bd7c4cecde424caf0f05
 *     four-channel:    8e4a41ebcc87183ca42f023409713ac2316f37c480bae00582536759543936fb
 *
 * For XRPL submission see submit_xrpl/SubmitXrpl.java (using xrpl4j).
 * For Bitcoin/OpenTimestamps see submit_bitcoin/SubmitBitcoin.java.
 */

import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.security.MessageDigest;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.*;

public class SecureFusionAnchor {

    private static final byte[] TEST_APP_SEED = new byte[32];
    static {
        for (int i = 0; i < 32; i++) TEST_APP_SEED[i] = (byte) i;
    }

    public static void main(String[] args) throws Exception {
        // Linked map: print order matches the test-vectors.json order.
        Map<String, String> expected = new LinkedHashMap<>();
        expected.put("single-channel-event.json", "e1b8a2206cf8c53754c392f8e6b7aad25972e6349a70bd7c4cecde424caf0f05");
        expected.put("four-channel-event.json",   "8e4a41ebcc87183ca42f023409713ac2316f37c480bae00582536759543936fb");

        System.out.println("SecureFusion v1.0 - Java anchor producer sample");
        System.out.println("=".repeat(60));

        Path examplesDir = findExamplesDir();
        boolean allOk = true;

        for (var e : expected.entrySet()) {
            Path manifestPath = examplesDir.resolve(e.getKey());
            String json = Files.readString(manifestPath, StandardCharsets.UTF_8);
            @SuppressWarnings("unchecked")
            Map<String, Object> manifest = (Map<String, Object>) TinyJson.parse(json);

            AnchorPayload payload = buildAnchorPayload(manifest, TEST_APP_SEED);
            boolean ok = payload.bundleHash.equals(e.getValue());
            allOk &= ok;

            System.out.println();
            System.out.println("  Manifest:    " + e.getKey());
            System.out.println("  bundleHash:  " + payload.bundleHash);
            System.out.println("  expected:    " + e.getValue());
            System.out.println("  match:       " + (ok ? "[OK]" : "[FAIL]"));
            System.out.println("  channels:    " + ((List<?>) manifest.get("channels")).size());
            System.out.println("  bundle:      " + payload.bundleBytes.length + " bytes");
            System.out.println("  signature:   " + toHexLower(payload.signature).substring(0, 32) + "...");
        }

        System.out.println();
        System.out.println(allOk ? "[OK] All test vectors match." : "[FAIL] One or more test vectors did not match.");
        System.exit(allOk ? 0 : 1);
    }

    private static Path findExamplesDir() {
        Path p = Path.of("").toAbsolutePath();
        for (int i = 0; i < 8 && p != null; i++, p = p.getParent()) {
            Path c = p.resolve("examples");
            if (Files.isDirectory(c)) return c;
        }
        throw new RuntimeException("Examples directory not found");
    }

    // ============================================================
    // Anchor builder.
    // ============================================================

    public record AnchorPayload(String bundleHash, byte[] bundleBytes, byte[] eventBytes,
                                byte[] signature, List<XrplMemo> memos) {}

    public record XrplMemo(String memoType, String memoFormat, String memoData) {}

    // SecureFusion v1: single-value enum, single source-code byte.
    private static final Map<String, Integer> SOURCE_CODES = Map.of(
        "fleetfusion", 1);

    public static AnchorPayload buildAnchorPayload(Map<String, Object> manifest, byte[] appSeed32) {
        byte[] eventBytes = CanonicalJson.serialise(manifest);
        String bundleHash = sha256Hex(eventBytes);

        byte[] bundleBytes = encodeBundleMemo(
            bundleHash,
            (String) manifest.get("eventId"),
            (String) manifest.get("ingestSource"),
            ((List<?>) manifest.get("channels")).size());

        byte[] sigInput = concat(bundleBytes, eventBytes);
        byte[] signature = appSeed32 == null ? new byte[64] : ed25519Sign(sigInput, appSeed32);

        List<XrplMemo> memos = List.of(
            new XrplMemo(hexUtf8("SF1.bundle"), hexUtf8("application/octet-stream"), toHexUpper(bundleBytes)),
            new XrplMemo(hexUtf8("SF1.event"),  hexUtf8("application/json"),         toHexUpper(eventBytes)),
            new XrplMemo(hexUtf8("SF1.sig"),    hexUtf8("application/octet-stream"), toHexUpper(signature)));

        return new AnchorPayload(bundleHash, bundleBytes, eventBytes, signature, memos);
    }

    public static byte[] encodeBundleMemo(String bundleHash, String eventId, String source, int channelCount) {
        if (bundleHash.length() != 64) throw new IllegalArgumentException("bundleHash must be 64 hex chars");
        Integer src = SOURCE_CODES.get(source);
        if (src == null) throw new IllegalArgumentException("Unknown ingestSource: " + source);
        if (channelCount < 1 || channelCount > 255) throw new IllegalArgumentException("channelCount out of range");

        byte[] out = new byte[50];
        System.arraycopy(fromHex(bundleHash), 0, out, 0, 32);
        System.arraycopy(toDotNetGuidBytes(eventId), 0, out, 32, 16);
        out[48] = src.byteValue();
        out[49] = (byte) channelCount;
        return out;
    }

    /**
     * Convert a UUID string to .NET Guid.ToByteArray() byte order (mixed-endian:
     * Data1/2/3 little-endian, Data4 verbatim). FleetFusion's
     * RippledXrpAnchorClient writes Guid bytes this way and the conformance
     * vectors are byte-stable on this convention.
     */
    static byte[] toDotNetGuidBytes(String uuidStr) {
        String hex = uuidStr.replace("-", "");
        if (hex.length() != 32) throw new IllegalArgumentException("bad uuid: " + uuidStr);
        byte[] raw = fromHex(hex);
        byte[] out = new byte[16];
        // Data1: 4 bytes little-endian
        out[0] = raw[3]; out[1] = raw[2]; out[2] = raw[1]; out[3] = raw[0];
        // Data2: 2 bytes little-endian
        out[4] = raw[5]; out[5] = raw[4];
        // Data3: 2 bytes little-endian
        out[6] = raw[7]; out[7] = raw[6];
        // Data4: 8 bytes verbatim
        System.arraycopy(raw, 8, out, 8, 8);
        return out;
    }

    /**
     * Sign with Ed25519 using BouncyCastle. The provider class is loaded by
     * reflection so the file compiles even if BC is absent on the classpath
     * (the runtime call will then fail with a clear error). Add bcprov-jdk18on
     * to the classpath to enable signing.
     */
    public static byte[] ed25519Sign(byte[] message, byte[] seed32) {
        try {
            Class<?> paramsClass = Class.forName("org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters");
            Class<?> signerClass = Class.forName("org.bouncycastle.crypto.signers.Ed25519Signer");
            Object params = paramsClass.getConstructor(byte[].class, int.class).newInstance(seed32, 0);
            Object signer = signerClass.getDeclaredConstructor().newInstance();
            signerClass.getMethod("init", boolean.class, Class.forName("org.bouncycastle.crypto.CipherParameters"))
                .invoke(signer, true, params);
            signerClass.getMethod("update", byte[].class, int.class, int.class)
                .invoke(signer, message, 0, message.length);
            return (byte[]) signerClass.getMethod("generateSignature").invoke(signer);
        } catch (ClassNotFoundException e) {
            throw new RuntimeException(
                "Ed25519 signing requires BouncyCastle on the classpath. " +
                "Add bcprov-jdk18on (e.g. via Maven's pom.xml or by dropping bcprov-jdk18on-1.78.jar into lib/).", e);
        } catch (Exception e) {
            throw new RuntimeException("BouncyCastle Ed25519 signing failed: " + e.getMessage(), e);
        }
    }

    // ============================================================
    // Canonical JSON.
    //
    // Hand-rolled because java.time.Instant.toString() drops trailing
    // zero-millisecond digits and json libraries (Jackson/Gson) take
    // configuration-dependent shortcuts that cannot be relied on across
    // versions. We emit timestamps with formatTimestamp() and walk the
    // tree manually with sorted keys.
    // ============================================================

    static final class CanonicalJson {
        static byte[] serialise(Object v) {
            StringBuilder sb = new StringBuilder();
            write(sb, v);
            return sb.toString().getBytes(StandardCharsets.UTF_8);
        }

        static void write(StringBuilder sb, Object v) {
            if (v == null) { sb.append("null"); return; }
            if (v instanceof Boolean b) { sb.append(b ? "true" : "false"); return; }
            if (v instanceof Long l) { sb.append(l); return; }
            if (v instanceof Integer i) { sb.append(i); return; }
            if (v instanceof Double d) {
                if (!Double.isFinite(d)) throw new IllegalArgumentException("Non-finite numbers not allowed");
                if (d == Math.floor(d) && Math.abs(d) < 1e16) sb.append((long) (double) d);
                else throw new IllegalArgumentException("Floats are forbidden in v1 manifests");
                return;
            }
            if (v instanceof Number n) { sb.append(n); return; }
            if (v instanceof String s) { writeString(sb, s); return; }
            if (v instanceof List<?> list) {
                sb.append('[');
                boolean first = true;
                for (Object item : list) { if (!first) sb.append(','); first = false; write(sb, item); }
                sb.append(']');
                return;
            }
            if (v instanceof Map<?, ?> map) {
                List<String> keys = new ArrayList<>();
                for (Object k : map.keySet()) keys.add((String) k);
                Collections.sort(keys);  // ordinal codepoint comparison (String.compareTo)
                sb.append('{');
                boolean first = true;
                for (String k : keys) {
                    if (!first) sb.append(',');
                    first = false;
                    writeString(sb, k);
                    sb.append(':');
                    write(sb, map.get(k));
                }
                sb.append('}');
                return;
            }
            throw new IllegalArgumentException("Cannot canonicalise " + v.getClass());
        }

        static void writeString(StringBuilder sb, String s) {
            sb.append('"');
            for (int i = 0; i < s.length(); i++) {
                char c = s.charAt(i);
                switch (c) {
                    case '"':  sb.append("\\\""); break;
                    case '\\': sb.append("\\\\"); break;
                    case '\b': sb.append("\\b");  break;
                    case '\f': sb.append("\\f");  break;
                    case '\n': sb.append("\\n");  break;
                    case '\r': sb.append("\\r");  break;
                    case '\t': sb.append("\\t");  break;
                    default:
                        if (c < 0x20) sb.append(String.format("\\u%04x", (int) c));
                        else sb.append(c);
                }
            }
            sb.append('"');
        }
    }

    /**
     * Format an ISO-8601 timestamp with explicit millisecond precision and
     * trailing Z. NEVER call Instant.toString() -- it strips trailing zeros
     * (e.g. "2026-04-30T12:00:00Z" instead of the spec-mandated
     * "2026-04-30T12:00:00.000Z"). Builds the string from epoch seconds + ms
     * via a fixed pattern.
     */
    public static String formatTimestamp(Instant instant) {
        return DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
            .withZone(ZoneOffset.UTC)
            .format(instant);
    }

    // ============================================================
    // Tiny JSON parser (so the sample has zero third-party deps for
    // reading the example manifests). For production, swap in your
    // existing JSON library.
    // ============================================================

    static final class TinyJson {
        private final String s;
        private int i;

        private TinyJson(String s) { this.s = s; }

        static Object parse(String json) {
            TinyJson t = new TinyJson(json);
            t.skipWs();
            Object v = t.value();
            t.skipWs();
            if (t.i != t.s.length()) throw new RuntimeException("Trailing content at " + t.i);
            return v;
        }

        Object value() {
            skipWs();
            char c = s.charAt(i);
            return switch (c) {
                case '{' -> obj();
                case '[' -> arr();
                case '"' -> str();
                case 't', 'f' -> bool();
                case 'n' -> nul();
                default -> num();
            };
        }

        Map<String, Object> obj() {
            Map<String, Object> m = new LinkedHashMap<>();
            i++; skipWs();
            if (s.charAt(i) == '}') { i++; return m; }
            while (true) {
                skipWs();
                String k = str();
                skipWs();
                if (s.charAt(i) != ':') throw new RuntimeException("Expected ':' at " + i);
                i++;
                m.put(k, value());
                skipWs();
                if (s.charAt(i) == ',') { i++; continue; }
                if (s.charAt(i) == '}') { i++; return m; }
                throw new RuntimeException("Expected ',' or '}' at " + i);
            }
        }

        List<Object> arr() {
            List<Object> a = new ArrayList<>();
            i++; skipWs();
            if (s.charAt(i) == ']') { i++; return a; }
            while (true) {
                a.add(value());
                skipWs();
                if (s.charAt(i) == ',') { i++; continue; }
                if (s.charAt(i) == ']') { i++; return a; }
                throw new RuntimeException("Expected ',' or ']' at " + i);
            }
        }

        String str() {
            if (s.charAt(i) != '"') throw new RuntimeException("Expected string at " + i);
            i++;
            StringBuilder sb = new StringBuilder();
            while (s.charAt(i) != '"') {
                char c = s.charAt(i++);
                if (c != '\\') { sb.append(c); continue; }
                char e = s.charAt(i++);
                switch (e) {
                    case '"', '\\', '/' -> sb.append(e);
                    case 'b' -> sb.append('\b');
                    case 'f' -> sb.append('\f');
                    case 'n' -> sb.append('\n');
                    case 'r' -> sb.append('\r');
                    case 't' -> sb.append('\t');
                    case 'u' -> {
                        sb.append((char) Integer.parseInt(s.substring(i, i + 4), 16));
                        i += 4;
                    }
                    default -> throw new RuntimeException("Bad escape \\" + e);
                }
            }
            i++;
            return sb.toString();
        }

        Boolean bool() {
            if (s.startsWith("true", i)) { i += 4; return Boolean.TRUE; }
            if (s.startsWith("false", i)) { i += 5; return Boolean.FALSE; }
            throw new RuntimeException("Bad boolean at " + i);
        }

        Object nul() {
            if (s.startsWith("null", i)) { i += 4; return null; }
            throw new RuntimeException("Bad null at " + i);
        }

        Number num() {
            int start = i;
            if (s.charAt(i) == '-') i++;
            while (i < s.length() && "0123456789.eE+-".indexOf(s.charAt(i)) >= 0) i++;
            String t = s.substring(start, i);
            if (t.contains(".") || t.contains("e") || t.contains("E")) return Double.parseDouble(t);
            return Long.parseLong(t);
        }

        void skipWs() {
            while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++;
        }
    }

    // ============================================================
    // Utilities.
    // ============================================================

    static byte[] concat(byte[] a, byte[] b) {
        byte[] r = new byte[a.length + b.length];
        System.arraycopy(a, 0, r, 0, a.length);
        System.arraycopy(b, 0, r, a.length, b.length);
        return r;
    }

    static String sha256Hex(byte[] data) {
        try {
            return toHexLower(MessageDigest.getInstance("SHA-256").digest(data));
        } catch (Exception e) { throw new RuntimeException(e); }
    }

    static String hexUtf8(String s) { return toHexUpper(s.getBytes(StandardCharsets.UTF_8)); }

    static String toHexUpper(byte[] b) {
        StringBuilder sb = new StringBuilder(b.length * 2);
        for (byte x : b) sb.append(String.format("%02X", x));
        return sb.toString();
    }

    static String toHexLower(byte[] b) {
        StringBuilder sb = new StringBuilder(b.length * 2);
        for (byte x : b) sb.append(String.format("%02x", x));
        return sb.toString();
    }

    static byte[] fromHex(String h) {
        byte[] out = new byte[h.length() / 2];
        for (int i = 0; i < h.length(); i += 2)
            out[i / 2] = (byte) ((Character.digit(h.charAt(i), 16) << 4) | Character.digit(h.charAt(i + 1), 16));
        return out;
    }
}
